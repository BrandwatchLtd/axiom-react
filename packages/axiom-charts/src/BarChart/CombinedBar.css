.ax-bar-chart__combined-bar_diff {
  position: absolute;
  top: 0;
  width: 100%;
}
