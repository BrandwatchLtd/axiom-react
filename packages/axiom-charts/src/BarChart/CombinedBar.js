import PropTypes from 'prop-types';
import React, { Component } from 'react';
import omit from 'lodash.omit';
import Bar from '../Bar/Bar';
import ChartContext from '../ChartContext/ChartContext';
import './CombinedBar.css';

export default class CombinedBar extends Component {
  static propTypes = {
    DropdownContext: PropTypes.func,
    TooltipContext: PropTypes.func,
    benchmarkValue: PropTypes.number.isRequired,
    color: PropTypes.string.isRequired,
    data: PropTypes.object.isRequired,
    label: PropTypes.node.isRequired,
    onDropdownClose: PropTypes.func.isRequired,
    onDropdownOpen: PropTypes.func.isRequired,
    percent: PropTypes.number.isRequired,
    value: PropTypes.number.isRequired,
  };

  render() {
    const {
      DropdownContext,
      TooltipContext,
      benchmarkValue,
      color,
      data,
      label,
      onDropdownClose,
      onDropdownOpen,
      percent,
      value,
      ...rest
    } = this.props;
    const isStretched = benchmarkValue > percent;

    const stripedBarWidth = isStretched ? benchmarkValue - percent : percent - benchmarkValue;
    const stripedBarStyle = {
      left: isStretched ? `${percent}%` : `${benchmarkValue}%`,
    };

    const stripedBarProps = omit(rest, [
      'onClick',
    ]);

    return (
      <React.Fragment>
        <ChartContext
            DropdownContext={ DropdownContext }
            TooltipContext={ TooltipContext }
            color={ color }
            data={ data }
            label={ label }
            onDropdownClose={ onDropdownClose }
            onDropdownOpen={ () => onDropdownOpen(color) }
            value={ value }>
          <Bar
              color={ color }
              percent={ percent }
              { ...rest } />
        </ChartContext>

        { isStretched && (<div className="ax-bar-chart__combined-bar-diff" style={ stripedBarStyle }>
          <Bar
              color={ color }
              fillMode="striped"
              percent={ stripedBarWidth }
              { ...stripedBarProps } />
        </div>) }
      </React.Fragment>
    );
  }
}
