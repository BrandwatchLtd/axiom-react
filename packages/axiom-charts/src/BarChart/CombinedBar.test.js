import React from 'react';
import renderer from 'react-test-renderer';
import CombinedBar from './CombinedBar';

function getComponent(newProps = {}) {
  const props = {
    DropdownContext: jest.fn(),
    TooltipContext: jest.fn(),
    onDropdownClose: jest.fn(),
    onDropdownOpen: jest.fn(),
    data: [],
    color: 'critical-mass',
    label: 'Possitve',
    value: 4321,
    ...newProps,
  };

  return renderer.create(
    <CombinedBar { ...props }
        color="giant-leap"
        direction="right" />
  );
}

describe('CombinedBar', () => {
  it('renders a single bar when benchmark value is less than current value', () => {
    const component = getComponent({
      benchmarkValue: 10,
      percent: 50,
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an additional striped bar when benchmark value is grather than current value', () => {
    const component = getComponent({
      benchmarkValue: 50,
      percent: 10,
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
