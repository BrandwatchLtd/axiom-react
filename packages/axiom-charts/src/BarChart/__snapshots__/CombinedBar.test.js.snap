// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CombinedBar renders a single bar when benchmark value is less than current value 1`] = `
<React.Fragment>
  <ChartContext
    DropdownContext={[MockFunction]}
    TooltipContext={[MockFunction]}
    color="giant-leap"
    data={Object {}}
    label="Possitve"
    onDropdownClose={[MockFunction]}
    onDropdownOpen={[Function]}
    value={4321}
  >
    <Bar
      color="giant-leap"
      direction="right"
      fillMode="solid"
      isFaded={false}
      isHidden={false}
      labelStrong={false}
      minSize="1rem"
      onClick={[MockFunction]}
      percent={50}
    />
  </ChartContext>
</React.Fragment>
`;

exports[`CombinedBar renders an additional striped bar when benchmark value is grather than current value 1`] = `
<React.Fragment>
  <ChartContext
    DropdownContext={[MockFunction]}
    TooltipContext={[MockFunction]}
    color="giant-leap"
    data={Object {}}
    label="Possitve"
    onDropdownClose={[MockFunction]}
    onDropdownOpen={[Function]}
    value={4321}
  >
    <Bar
      color="giant-leap"
      direction="right"
      fillMode="solid"
      isFaded={false}
      isHidden={false}
      labelStrong={false}
      minSize="1rem"
      onClick={[MockFunction]}
      percent={10}
    />
  </ChartContext>
  <div
    className="ax-bar-chart__combined-bar-diff"
    style={
      Object {
        "left": "10%",
      }
    }
  >
    <ChartContext
      TooltipContext={[MockFunction]}
      color="giant-leap"
      data={Object {}}
      label="Possitve"
      value={-4271}
    >
      <Bar
        color="giant-leap"
        direction="right"
        fillMode="striped"
        isFaded={false}
        isHidden={false}
        labelStrong={false}
        minSize="1rem"
        percent={40}
      />
    </ChartContext>
  </div>
</React.Fragment>
`;
